tests/stdlib/tlists.nim

```diff
@@ -82,87 +82,60 @@ block tlistsToString:
-  type Foo = ref object
-    x: int
-  var f = Foo(x: 1)
-  let a = [f, f].toSinglyLinkedList
-  f.x = 42
-  assert a.head.value == f
-  assert a.head.next.value == f
+  template testCommon(toList) =
+    doAssert array[0, int].default.toList.copy.toSeq == []
+    doAssert [1].toList.copy.toSeq == [1]
+    doAssert [1, 2].toList.copy.toSeq == [1, 2]
+    doAssert [1, 2, 3].toList.copy.toSeq == [1, 2, 3]
+    type Foo = ref object
+      x: int
+    var f = Foo(x: 1)
+    let a = [f, f].toList
+    f.x = 42
+    assert a.head.value == f
+    assert a.head.next.value == f
+  testCommon(toSinglyLinkedList)
+  testCommon(toDoublyLinkedList)
 
 block add:
-  block: # SinglyLinkedList
+  template testCommon(initList, toList) =
     block:
       var
-        l0 = initSinglyLinkedList[int]()
-        l1 = [1].toSinglyLinkedList
-        l2 = [2, 3].toSinglyLinkedList
-        l3 = [4, 5, 6].toSinglyLinkedList
+        l0 = initList[int]()
+        l1 = [1].toList
+        l2 = [2, 3].toList
+        l3 = [4, 5, 6].toList
         l4 = l3.copy
         l5 = l3.copy
-      l0.add(l3)
-      l1.add(l4)
-      l2.add(l5)
+      l0.add l3
+      l1.add l4
+      l2.add l5
       doAssert l0.toSeq == [4, 5, 6]
       doAssert l1.toSeq == [1, 4, 5, 6]
       doAssert l2.toSeq == [2, 3, 4, 5, 6]
       doAssert l3.toSeq == []
-      l2.add(l3) # re-adding l3 that was destroyed is now a no-op
+      l2.add l3 # re-adding l3 that was destroyed is now a no-op
       doAssert l2.toSeq == [2, 3, 4, 5, 6]
       doAssert l3.toSeq == []
     block:
       var
-        l0 = initSinglyLinkedList[int]()
-        l1 = [1].toSinglyLinkedList
-        l2 = [2, 3].toSinglyLinkedList
-        l3 = [4, 5, 6].toSinglyLinkedList
-      l3.add(l0)
-      l2.add(l1)
-      doAssert l3.toSeq == [4, 5, 6]
-      doAssert l2.toSeq == [2, 3, 1]
-  block: # DoublyLinkedList
-    block:
-      var
-        l0 = initDoublyLinkedList[int]()
-        l1 = [1].toDoublyLinkedList
-        l2 = [2, 3].toDoublyLinkedList
-        l3 = [4, 5, 6].toDoublyLinkedList
-        l4 = l3.copy
-        l5 = l3.copy
-      l0.add(l3)
-      l1.add(l4)
-      l2.add(l5)
-      doAssert l0.toSeq == [4, 5, 6]
-      doAssert l1.toSeq == [1, 4, 5, 6]
-      doAssert l2.toSeq == [2, 3, 4, 5, 6]
-      doAssert l3.toSeq == []
-      l2.add(l3) # re-adding l3 that was destroyed is now a no-op
-      doAssert l2.toSeq == [2, 3, 4, 5, 6]
-      doAssert l3.toSeq == []
-    block:
-      var
-        l0 = initDoublyLinkedList[int]()
-        l1 = [1].toDoublyLinkedList
-        l2 = [2, 3].toDoublyLinkedList
-        l3 = [4, 5, 6].toDoublyLinkedList
-      l3.add(l0)
-      l2.add(l1)
+        l0 = initList[int]()
+        l1 = [1].toList
+        l2 = [2, 3].toList
+        l3 = [4, 5, 6].toList
+      l3.add l0
+      l2.add l1
       doAssert l3.toSeq == [4, 5, 6]
       doAssert l2.toSeq == [2, 3, 1]
+  testCommon(initSinglyLinkedList, toSinglyLinkedList)
```

Comment:
you'd have less boilerplate with a single `testCommon`, and sub-blocks for each common proc

```
block: # SinglyLinkedList, DoublyLinkedList
  template testCommon(initList, toList) =
    block copy: ...
    block add: ...
    ...
  testCommon(initSinglyLinkedList, toSinglyLinkedList)
```

Reply:
I also considered that, but I assumed that a single top-level block for each tested procedure was the standard :)